pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -300,19 +317,75 @@ func (p *PriorityQueue) Run() {
 	go wait.Until(p.flushUnschedulablePodsLeftover, 30*time.Second, p.stop)
 }
 
+// PreEnqueue firstly runs registered EnqueuePlugins, and returns true if all plugins run
+// successfully; otherwise add or update the pod in `unschedulablePods`, and returns false.
+func (p *PriorityQueue) PreEnqueue(pInfo *framework.QueuedPodInfo) bool {
+	success := p.runEnqueuePlugins(context.Background(), pInfo)
+	if !success {
+		// Add the Pod to unschedulablePods if it's not passing EnqueuePlugins.
+		p.unschedulablePods.addOrUpdate(pInfo)
+	}
+	pInfo.Gated = !success
+	return success
+}
+
+// runEnqueuePlugins iterates PreEnqueue function in each registered EnqueuePlugin.
+// It returns true if all PreEnqueue function run successfully; otherwise returns false
+// upon the first failure.
+// Note: we need to associate the failed plugin to `pInfo`, so that the pod can be moved back
+// to activeQ by related cluster event.
+func (p *PriorityQueue) runEnqueuePlugins(ctx context.Context, pInfo *framework.QueuedPodInfo) bool {
```

Comment:
should we move this function to the framework just like all run*Plugins functions?

Reply:
nope, `frameworkImpl` doesn't hold scheudlingqueue, and hence cannot perform `runEnqueuePlugins`.